### PR TITLE
Create 4 classes. library_terminal is script for run.

### PR DESCRIPTION
**Pull request by andreykraveckiy**
In the next step to add files with data and read this  files in the library-object.

### CODE REVIEW

File: library.rb [hunk 1]
```diff
@@ -1,3 +1,120 @@
...
+    @readers = []
+    @orders = []
+    puts 'Use :load method for load library from file.'
+  end
+
+  def load
+    load_from_file('Author', @authors, 'author.txt')
+    load_from_file('Book', @books, 'book.txt')
+    load_from_file('Reader', @readers, 'reader.txt')
+    load_from_file('Order', @orders, 'order.txt')
```
Comment (yamnich) [inline library.rb]: 1. we can use only string author, book... as parameter and produce class name, instance variable name and file name using it
2. we can use array of 'author', 'book' ... and call load_from_file in the loop

File: library.rb [hunk 2]
```diff
@@ -1,3 +1,120 @@
...
+  end
+
+  def save
+    save_to_file(@authors, 'author.txt')
```
Comment (yamnich) [inline library.rb]: the same

File: library.rb [hunk 3]
```diff
@@ -1,3 +1,120 @@
...
+    end
+    new_class_instance("Order", @orders, params)
+  end  
+
+  def best_reader
+    readers = @orders.map(&:reader)
+    uniq = readers.uniq
+    good_name = uniq.first
+    uniq.each do |name|
+      good_name = name if readers.count(good_name) < readers.count(name)
```
Comment (yamnich) [inline library.rb]: why do we need it here?

File: library.rb [hunk 4]
```diff
@@ -1,3 +1,120 @@
...
+    end
+    answer = [good_name]
+    uniq.each do |name|
+      answer << name if readers.count(good_name) == readers.count(name) && good_name != name
```
Comment (yamnich) [inline library.rb]: why do we need it here?

File: library.rb [hunk 5]
```diff
@@ -1,3 +1,120 @@
...
+    uniq = books.uniq
+    good_title = uniq.first
+    uniq.each do |title|
+      good_title = title if books.count(good_title) < books.count(title)
+    end
+    answer = [good_title]
+    uniq.each do |title|
+      answer << title if books.count(good_title) == books.count(title) && good_title != title
+    end
+    answer
```
Comment (yamnich) [inline library.rb]: very long and complex method, logic is not clear enough

File: library.rb [hunk 6]
```diff
@@ -1,3 +1,120 @@
...
+  end
+
+  def the_3_most_popular
+    books = @orders.map(&:book)
+    uniq = books.uniq
```
Comment (yamnich) [inline library.rb]: what does uniq mean here?

File: library.rb [hunk 7]
```diff
@@ -1,3 +1,120 @@
...
+    uniq.sort! { |x,y| books.count(y)<=>books.count(x) }
+    "#{uniq.first}-#{books.count(uniq.first)}; #{uniq[1]}-#{books.count(uniq[1])}; #{uniq[3]}-#{books.count(uniq[2])}"
```
Comment (yamnich) [inline library.rb]: this method should return count of unique readers who take one of 3 popular book

File: library_terminal.rb [hunk 8]
```diff
@@ -0,0 +1,122 @@
...
+|    lw - sawe library to files     |
+|    eo - extra orders              |
+|    br - the best reader           |
+|    pb - the most popuar book      |
+|    3mp - the 3 most popuar books  |
+|    q - EXIT                       |
++-----------------------------------+
+  }
+
+$lib = Library.new
```
Comment (yamnich) [inline library_terminal.rb]: why do you use $lib here?

File: library.rb [hunk 9]
```diff
@@ -1,3 +1,119 @@
...
+    end
+  end
+
+# for interface
+  def new_author(params)
+    new_class_instance("Author", @authors, params)
+  end
+
+  def new_book(params)
+    if !@authors.map(&:name).index(params.last)
```
Comment (yamnich) [inline library.rb]: unless

File: library.rb [hunk 10]
```diff
@@ -1,3 +1,119 @@
...
+    end
+    new_class_instance("Book", @books, params)
+  end
+
+  def new_reader(params)
+    new_class_instance("Reader", @readers, params)
+  end
+
+  def new_order(params)
+    if !@books.map(&:title).index(params.first)
```
Comment (yamnich) [inline library.rb]: unless

File: library.rb [hunk 11]
```diff
@@ -1,3 +1,119 @@
...
+      puts "Library does not have the book #{params.first}"
+      return
+    end
+    if !@readers.map(&:name).index(params[1])
```
Comment (yamnich) [inline library.rb]: unless

File: library.rb [hunk 12]
```diff
@@ -1,3 +1,119 @@
...
+    end
+    new_class_instance("Order", params)
+  end  
+
+  def best_reader
+    readers = @orders.map(&:reader)
+    uniq_readers = readers.uniq
+    uniq_readers.sort! { |x,y| readers.count(y) <=> readers.count(x) }
+    # in case when we have few readers with equel quantity read books
+    uniq_readers.select { |e| readers.count(e) == readers.count(uniq_readers.first) }
```
Comment (yamnich) [inline library.rb]: Please give an example of order. I think that you do a lot of extra things. I will continue review after.

Comment (andreykraveckiy) [reply library.rb]: Add in order.txt

File: library.rb [hunk 13]
```diff
@@ -1,3 +1,129 @@
...
+  def best_reader
+    readers = @orders.map(&:reader)
+    uniq_readers = readers.uniq
+    uniq_readers.sort! { |x,y| readers.count(y) <=> readers.count(x) }
+    # in case when we have few readers with equel quantity read books
+    # example in the order txt file
+    # Best reader is(are):
+    # Andrey Kraveckiy
+    # Sergey Evtushenko
+    uniq_readers.select { |e| readers.count(e) == readers.count(uniq_readers.first) }
```
Comment (andreykraveckiy) [inline library.rb]: Tatiana, here you had asked me to get an example. I've made this in order txt (see below). In Comment you will see unswers.

Comment (yamnich) [reply library.rb]: According to your example, all you need to do:
1. group orders by reader - that you have reader => arrays of their orders
2. sort result of grouping by arrays of their orders.count
3. Take first/last item and reader from it

File: library.rb [hunk 14]
```diff
@@ -1,3 +1,129 @@
...
+    uniq_books = books.uniq
+    uniq_books.sort! { |x,y| books.count(y) <=> books.count(x) }
+    # in case when we have few books with equel quantity in orders
+    # example in the order txt file
+    # Popular book is(are):
+    # Ruslan and Liudmila
+    # People like a goods
+    # About person, life and nature
+
+    uniq_books.select { |e| books.count(e) == books.count(uniq_books.first) }
```
Comment (yamnich) [inline library.rb]: the same here

File: library.rb [hunk 15]
```diff
@@ -1,3 +1,129 @@
...
+
+  def the_3_most_popular
+    books = @orders.map(&:book)
+    uniq_books = books.uniq
+    uniq_books.sort! { |x,y| books.count(y) <=> books.count(x) }
+    answer = []
+    uniq_books.first(3).each do |book|
+      answer += @orders.map { |ord| ord.reader if ord.book == book }
+    end
+    answer.compact.uniq.length
```
Comment (yamnich) [inline library.rb]: 1. You can use modified method popular_book with parameter 3 to get popular books. Then select orders with these books, group them by reader and count 
2. You can group orders by book, sort by their orders count, get orders, group by reader, count

File: library.rb [hunk 16]
```diff
@@ -1,3 +1,117 @@
...
+      return
+    end
+    unless @readers.map(&:name).index(params[1])
+      puts "#{params[1]} wasn't registered at this library"
+      return
+    end
+    new_class_instance("Order", params)
+  end  
+
+  def best_reader
```
Comment (andreykraveckiy) [inline library.rb]: rewrite this

File: library.rb [hunk 17]
```diff
@@ -1,3 +1,117 @@
...
+    #readers = @orders.map(&:reader)
+    #uniq_readers = readers.uniq
+    #uniq_readers.sort! { |x,y| readers.count(y) <=> readers.count(x) }
+    # Best reader is(are):
+    # Andrey Kraveckiy
+    # Sergey Evtushenko
+    sort_readers.select { |e| group_by_reader[e].count == group_by_reader[sort_readers.first].count }
+  end
+
+  def popular_book(counter = nil)
```
Comment (andreykraveckiy) [inline library.rb:62]: rewrite method below

File: library.rb [hunk 18]
```diff
@@ -1,3 +1,117 @@
...
+    group_by_book= @orders.group_by { |e| e.book }
+    sort_books = group_by_book.keys.sort { |x,y| group_by_book[y].count <=> group_by_book[x].count}
+    return sort_books.first(counter) if counter
+    sort_books.select { |e| group_by_book[e].count == group_by_book[sort_books.first].count }
+  end
+
+  def the_3_most_popular
+    books = popular_book(3)
+    @orders.select { |e| books.include?(e.book) }.map(&:reader).uniq.count
```
Comment (andreykraveckiy) [inline library.rb:68]: rewrite this

**Review by andreykraveckiy — COMMENTED** [review]
Rewrite 3 methods

File: library.rb [hunk 19]
```diff
@@ -1,3 +1,110 @@
...
+    unless @readers.map(&:name).index(params[1])
+      puts "#{params[1]} wasn't registered at this library"
+      return
+    end
+    new_class_instance("Order", params)
+  end  
+
+  def best_reader
+    group_by_reader = @orders.group_by { |e| e.reader }
+    sort_readers = group_by_reader.keys.sort { |x,y| group_by_reader[y].count <=> group_by_reader[x].count }
```
Comment (yamnich) [inline library.rb]: 1. Why do you sort keys? You can just do something like group_by_reader.sort_by { |reader, orders| orders.size }.last or first
2. Why do you need sort_readers.select { |e| group_by_reader[e].count == group_by_reader[sort_readers.first].count } ?

You increase complexity using additional cycles and extra variables.

File: library.rb [hunk 20]
```diff
@@ -1,3 +1,110 @@
...
+    sort_readers.select { |e| group_by_reader[e].count == group_by_reader[sort_readers.first].count }
+  end
+
+  def popular_book(counter = nil)
+    group_by_book= @orders.group_by { |e| e.book }
+    sort_books = group_by_book.keys.sort { |x,y| group_by_book[y].count <=> group_by_book[x].count }
+    return sort_books.first(counter) if counter
+    sort_books.select { |e| group_by_book[e].count == group_by_book[sort_books.first].count }
```
Comment (yamnich) [inline library.rb]: 1. The same
2. These two methods are very similar. I think you can unify them.

File: library.rb (hunk 18 above)
Comment (yamnich) [reply library.rb:68]: very good!

File: library.rb [hunk 21]
```diff
@@ -85,6 +80,13 @@ def load_from_file(class_name)
       end
     end
 
+    def popular_in_order_by(field, counter = nil)
+      grouped_by = @orders.group_by { |e| e.send(field) }
+      sort_fields = grouped_by.keys.sort { |x,y| grouped_by[y].count <=> grouped_by[x].count }
+      return sort_fields.first(counter) if counter
+      sort_fields.select { |e| grouped_by[e].count == grouped_by[sort_fields.first].count }
```
Comment (andreykraveckiy) [inline library.rb]: 1.Why I use select here? Really, it is more logically to take first element as the best reader or the most popular book. In this case evident method is the next: 
sort_fields = grouped_by.keys.sort { |x,y| grouped_by[y].count <=> grouped_by[x].count }.first
BUT in  real library we can have 2 or 3 or n readers who has the same quantity of read books. In this case who is the best? Is the first of this men, second or reader with number n? I think all this people achieve the first position. And the select allow us to get names of thise really bookworms.
2.Why do I sort keys? group_by returns the Hash, where key is the field(book or reader) and value is array of orders, in which this field exists. When I take keys I get all names of users or books as array of unique values.

File: library.rb (hunk 18 above)
Comment (andreykraveckiy) [reply library.rb:68]: Thank you! =)

File: library.rb (hunk 21 above)
Comment (yamnich) [reply library.rb]: 1. According to task it should 1, but ok, let left it as it is
2. Yes, I know. That's why it is enough to use.sort_by { |name, orders| orders.size } or even max_by

**Review by yamnich — APPROVED** [review]

**Pull request merged**